### ISSUE TITLE
Support for float16 / mixed-precision computation

### ISSUE THREAD

**Issue by romainvo** [astra-toolbox/astra-toolbox#623]
Hi everyone! I would like to start by saying that the library is super cool. I use it almost daily for my work with the [deepinv](https://deepinv.github.io/deepinv/api/stubs/deepinv.physics.TomographyWithAstra.html#deepinv.physics.TomographyWithAstra) wrapper, and it works great :)!

I am curious to know if there are any plans on your side to support float16 computation (we may have discussed this already, @askorikov, but I forgot)? I remember that [torch-radon](https://github.com/matteo-ronchetti/torch-radon) used to have this kind of feature, and the speed difference compared to float32 was huge. 

This would be super useful for many users at the interface of deep learning and computational imaging (cc @Andrewwango) , e.g. for mixed/reduced precision training of unrolled-like models, or just for bare acceleration at inference.

**Comment by Andrewwango** [comment]
+1 thanks for developing this great library!

**Comment by wjp** [comment]
Hi,

Thank you for your interest. It's great to hear you're finding astra useful.

Float16 support is definitely something we'd consider, and in fact we've already been putting some of the infrastructure for it in place. Could you maybe expand a bit on what functionality you'd exactly like to have?

I.e., 2D and/or 3D? The BP/FP operations, and/or FBP/FDK reconstruction (if yes, including the filter?), and/or our SIRT/CGLS/... iterative solvers? Using specifically the `torch.float16`/`numpy.float16` data type? Are you envisioning mixing float16 and float32 buffers?

Sorry for the many questions; I hope to get a feeling for what exactly would be of interest, so we can focus our initial efforts on what would be most immediately useful, and then expand later.

**Comment by romainvo** [comment]
Thanks for answering so quickly :)! I'll try to answer as clearly as possible. 

1. At the moment we use the `GPULink` feature of astra (like tomosipo), to apply BP/FP operations on `torch.Tensor` stored on a cuda device (we re-implemented the filtering and solvers in torch for convenience so this is fine).  From my understanding, the `GPULink` feature is only supported for 3D geometries, but this is fine at the moment for us, we define a 2D geometry by creating a flat 3D one.

2. Concerning the data-type, yes, it would be `torch.float16`. `torch` as an [automatic mixed-precision](https://docs.pytorch.org/docs/stable/notes/amp_examples.html) feature but I think this is outside this scope. To put it simply, the ideal case would being able to pass a `torch.float16` Tensor (no casting under the astra hood for speed, etc..) and retrieve on the output a `torch.float16`.
   - What we do in practice is we treat all "astra" computations as a black-box, e.g we pass a `torch.float32` tensor, and retrieve a tensor of the same dtype/device. The astra computations are wrapped in a custom `torch.autograd.Function` so that computation of the gradients writh respect to the inputs is seamless for the user.
   - https://github.com/deepinv/deepinv/blob/b9b60ac64b1ac04df092c6ab6db1c11dd5e80a6d/deepinv/physics/functional/astra.py#L231 (how we use astra)

Here is a snippet of what kind of computation we do. Below I do the casting to `torch.float32` before passing the Tensor to astra but ideally I would not, and all computations would happen in `float16`.
```python
import torch
import deepinv as dinv
torch.autograd.set_detect_anomaly(True)

device = "cuda"
model = torch.nn.Conv2d(1, 1, 3, padding=1).to(torch.float16).to(device)
optim = torch.optim.Adam(model.parameters(), lr=1e-4)

# uses astra backend
physics = dinv.physics.TomographyWithAstra(
    img_size=(32,32),
    angles=32,
    n_detector_pixels=32
)

for i in range(10):
    # Some input in float16
    x = torch.randn(1, 1, 32, 32, device=device, dtype=torch.float16)
    # Some measurements (in float32 for astra)
    y = torch.randn(1, 1, 32, 32, device=device, dtype=torch.float32)

    # Unrolled-like computation, where the reconstruction steps composes
    # network steps and ct steps
    out = model(x)
    
    # gradient-step, out is casted to float32 because astra only supports float32
    out = out - 1e-4 * physics.A_adjoint(physics.A(out.to(torch.float32)) - y)

    # cast back to float16 to speed up model step
    out = out.to(torch.float16)
    
    out = model(out)

    # Some loss then backprop to train model
    loss = torch.mean(out**2)
    loss.backward()
    optim.step()
    optim.zero_grad()
```

Other use-case in mind @Andrewwango ?

**Comment by Andrewwango** [comment]
Thanks for your response @wjp !

Yes, I think `bfloat16` and/or `float16` support of the fundamental FP/BP operator for 3D applications is the biggest need, where memory is often limited. On a side note we create geometries with `creators.create_proj_geom`, `creators.create_vol_geom` so it would be nice to have support there, otherwise we can always cast things offline!

**Comment by wjp** [comment]
Thank you for the details. @askorikov and I have discussed this internally, and we'll explore the possibilities in the near future.
